src/optable.ts

```diff
@@ -520,27 +521,45 @@ optable[opcodes.LOAD_GLOBAL] = function(f: Py_FrameObject) {
 
 optable[opcodes.LOAD_DEREF] = function(f: Py_FrameObject) {
     var i = f.readArg();
-
-    var numCellvars = f.codeObj.cellvars.length;
-    if (i < numCellvars) {
-        var name = f.codeObj.cellvars[i].toString();
-        f.push(f.locals[name]);
-    } else {
-        f.push(f.env[i - numCellvars]);
+    var cell = <Py_Cell> f.env[i];
+    
+    if (cell.ob_ref === null) {
+        var name;
+        var numCellvars = f.codeObj.cellvars.length;
+        if (i < numCellvars) {
+            name = f.codeObj.cellvars[i].toString();
+
+        } else {
+            name = f.codeObj.freevars[i - numCellvars].toString();
+        }
+        cell.ob_ref = f.locals[name];
     }
+    f.push(cell.ob_ref);
+
+}
+
+optable[opcodes.STORE_DEREF] = function(f: Py_FrameObject) {
+    var i = f.readArg();
+    var obj = f.pop();
+    f.env[i].ob_ref = obj;
 }
 
 optable[opcodes.LOAD_CLOSURE] = function(f: Py_FrameObject) {
     var i = f.readArg();
-    // Pushes a reference to the cell contained in slot i of the cell and free variable storage.
-    var numCellvars = f.codeObj.cellvars.length;
-    var name: primitives.Py_Str;
-    if (i < numCellvars) {
-      name = f.codeObj.cellvars[i];
-    } else {
-      name = f.codeObj.freevars[i - numCellvars];
+    // Pushes a reference to the cell contained in slot i of the cell and free variable storage
+    var cell = <Py_Cell> f.env[i];
```

Comment:
Ya, but I am downcasting from IPy_Object. The function's closure is currently a `Py_Tuple` (CPython has it that way), though it might as well be a `Py_Cell[]`.